### PR TITLE
Add isQuery() to vector layer and data provider

### PR DESCRIPTION
**Pull request by elpaso**
The method returns TRUE if the layer is a query/sql layer.

This allows to selectively show the "Update SQL Layer..." menu
entry in the legend.

Partial fix for #45796

### CODE REVIEW

File: python/core/auto_generated/vector/qgsvectordataprovider.sip.in [hunk 1]
```diff
@@ -128,6 +128,19 @@ Returns ``True`` if the layer does not contain any feature.
 .. versionadded:: 3.4
 %End
 
+    virtual bool isQuery() const;
```
Comment (nyalldawson) [inline python/core/auto_generated/vector/qgsvectordataprovider.sip.in:131]: I wonder if it'd be a bit more flexible to instead have a "layerType" method which returns an enum instead (initially with Table/Query values).

**Comment by kannes** [comment]
Something like this would also be very useful to show that they are not editable in the mouse-over of the layer name or layer properties. Took me a moment recently to realise why a PostGIS layer wasn't editable until I realised it was a query layer. :)

File: python/core/auto_generated/vector/qgsvectordataprovider.sip.in (hunk 1 above)
Comment (elpaso) [reply python/core/auto_generated/vector/qgsvectordataprovider.sip.in]: Done

File: src/core/qgis.h [hunk 2]
```diff
@@ -135,6 +135,20 @@ class CORE_EXPORT Qgis
...
+     *
+     * \since QGIS 3.24
+     */
+    enum class VectorLayerTypeFlag SIP_MONKEYPATCH_SCOPEENUM_UNNEST( Qgis, VectorLayerTypeFlag ) : int
+      {
+      SqlQuery = 1 << 0 //!< SQL query layer
+    };
+    Q_ENUM( VectorLayerTypeFlag )
+    //! Vector layer type flags
+    Q_DECLARE_FLAGS( VectorLayerTypeFlags, VectorLayerTypeFlag ) SIP_MONKEYPATCH_FLAGS_UNNEST( QgsVectorLayer, VectorLayerTypeFlags )
```
Comment (nyalldawson) [inline src/core/qgis.h:149]:
```suggestion
    Q_DECLARE_FLAGS( VectorLayerTypeFlags, VectorLayerTypeFlag );
```

This is only needed for enums which are moved

File: src/core/qgis.h [hunk 3]
```diff
@@ -135,6 +135,20 @@ class CORE_EXPORT Qgis
     };
     Q_ENUM( DataType )
 
+    /**
+     * Vector layer type flags.
+     *
+     * \since QGIS 3.24
+     */
+    enum class VectorLayerTypeFlag SIP_MONKEYPATCH_SCOPEENUM_UNNEST( Qgis, VectorLayerTypeFlag ) : int
```
Comment (nyalldawson) [inline src/core/qgis.h:143]:
```suggestion
    enum class VectorLayerTypeFlag : int
```

File: python/core/auto_generated/vector/qgsvectorlayer.sip.in [hunk 4]
```diff
@@ -451,6 +451,31 @@ This is obtained from the data provider and does not follow any standard.
...
+   is set.
+
+.. seealso:: :py:func:`vectorLayerTypeFlags`
+
+.. versionadded:: 3.24
+%End
+
+    Qgis::VectorLayerTypeFlags vectorLayerTypeFlags() const;
+%Docstring
+Returns ``True`` if the layer is a query (SQL) layer.
```
Comment (nyalldawson) [inline python/core/auto_generated/vector/qgsvectorlayer.sip.in:471]: This needs updating

File: src/core/qgis.h [hunk 5]
```diff
@@ -135,6 +135,20 @@ class CORE_EXPORT Qgis
...
+     *
+     * \since QGIS 3.24
+     */
+    enum class VectorLayerTypeFlag : int
+    {
+      SqlQuery = 1 << 0 //!< SQL query layer
+    };
+    Q_ENUM( VectorLayerTypeFlag )
+    //! Vector layer type flags
+    Q_DECLARE_FLAGS( VectorLayerTypeFlags, VectorLayerTypeFlag )
```
Comment (nyalldawson) [inline src/core/qgis.h:149]: Also needs Q_DECLARE_OPERATORS_FOR_FLAGS added for the new flags

**Review by nyalldawson — APPROVED** [review]

**Comment by jschultz** [comment]
If this comment gets noticed I'd like to signal that downstream this change seems to have caused things to stop working. Noone is interested on `qgis-users` so I'm climbing up the food chain. I think I can see where the problem might lie. The function isQuery does not seem to capture all SQL layers. For instance a straightforward PostGIS layer with the URI:
`dbname='xxx' port=5432 user='xxx' sslmode=disable key='id' srid=4283 type=MultiSurface checkPrimaryKeyUnicity='0' table="public"."xxxx"(geom)`
which previously could have its SQL updated, is now no longer able.

**Comment by elpaso** [comment]
> 

This is not what we call a "query layer". A query layer is a layer that originates from the results of a complex query, you can create one in DB manager or from the browser's execute SQL command menu over a DB connection.

**Comment by jschultz** [comment]
Well maybe it was an unintended feature but it worked great and was very useful for tweaking PostGIS layers.

**Comment by elpaso** [comment]
> Well maybe it was an unintended feature but it worked great and was very useful for tweaking PostGIS layers. Actually more than tweaking - you could basically make a layer out of any SQL query. Brilliant!

You can still do that from DB manager and from the browser by executing an arbitrary query and creating a layer from the results.

IIRC previous versions allowed to silently turn any DB layer into a query layer but there was no way back, query layers are read-only and have other limitations, now this possibility no longer exists but as I mentioned there are multiple ways to create a query layer and once you have created such a layer its SQL command can be updated.

**Comment by jschultz** [comment]
OK I can understand that. Thanks for the quick answers!

**Comment by jschultz** [comment]
Actually while I am here how will the isQuery criterion handle queries with Common Table Expressions, aka 'WITH' clauses?

**Pull request merged**